File: ones-gfx-sdk-python/ones_gfx/resources/tenants.py
from __future__ import annotations

from typing import Any
from urllib.parse import urlparse

from ..enums import OperationMode
from ..models import Operation, Tenant
from ..transport import Transport

_OPERATION_KEYS = {"operationId", "status"}


class TenantsResource:
    def __init__(self, transport: Transport):
        self._transport = transport

    def list(self, fabric_name: str) -> list[Tenant]:
        _require(fabric_name, "fabric_name")
        body = self._transport.get(f"fabrics/{fabric_name}/tenants")
        items = body.get("tenants", []) if isinstance(body, dict) else []
        return [Tenant.from_api(item) for item in items]

    def get(self, fabric_name: str, name: str) -> Tenant:
        _require(fabric_name, "fabric_name")
        _require(name, "name")
        body = self._transport.get(f"fabrics/{fabric_name}/tenants/{name}")
        tenant_payload = body.get("tenant", body) if isinstance(body, dict) else body
        return Tenant.from_api(tenant_payload)

    def available_servers(self, fabric_name: str) -> list[str]:
        _require(fabric_name, "fabric_name")
        body = self._transport.get(f"fabrics/{fabric_name}/available_servers")
        return body.get("availableGPUs", []) if isinstance(body, dict) else []

    def create(
        self,
        fabric_name: str,
        name: str,
        description: str,
        max_gpus_allowed: int,
        *,
        shared: bool = False,
        mode: OperationMode = OperationMode.SYNCHRONOUS,
        webhook_url: str | None = None,
        webhook_events: list[str] | None = None,
    ) -> Tenant | Operation:
        _require(fabric_name, "fabric_name")
        _require(name, "name")
        _validate_max_gpus(max_gpus_allowed)
        body = {"tenantName": name, "description": description or "", "maxGpusAllowed": max_gpus_allowed, "shared": shared}
        _attach_webhook_fields(body, mode, webhook_url, webhook_events)
        result = self._transport.post(f"fabrics/{fabric_name}/tenants", json_body=body, mode=mode)
        if _looks_like_operation(result):
            return Operation.from_submission(result)
        tenant_payload = result.get("tenant", result) if isinstance(result, dict) else result
        return Tenant.from_api(tenant_payload)

    def delete(
        self,
        fabric_name: str,
        name: str,
        *,
        mode: OperationMode = OperationMode.SYNCHRONOUS,
        webhook_url: str | None = None,
        webhook_events: list[str] | None = None,
    ) -> Operation | None:
        _require(fabric_name, "fabric_name")
        _require(name, "name")
        body = None
        if mode is OperationMode.ASYNC_WEBHOOK:
            body = {}
            _attach_webhook_fields(body, mode, webhook_url, webhook_events)
        if body is None:
            result = self._transport.delete(f"fabrics/{fabric_name}/tenants/{name}", mode=mode)
        else:
            result = self._transport._request("DELETE", f"fabrics/{fabric_name}/tenants/{name}", json_body=body, mode=mode)  # noqa: SLF001
        if _looks_like_operation(result):
            return Operation.from_submission(result)
        return None

    def allocate_gpus(self, fabric_name: str, name: str, servers: list[str | dict], *, mode: OperationMode = OperationMode.SYNCHRONOUS, webhook_url: str | None = None, webhook_events: list[str] | None = None, timeout: int | None = None) -> Operation | None:
        return self._gpu_update(fabric_name, name, "ADD", servers, mode=mode, webhook_url=webhook_url, webhook_events=webhook_events, timeout=timeout)

    def deallocate_gpus(self, fabric_name: str, name: str, servers: list[str | dict], *, mode: OperationMode = OperationMode.SYNCHRONOUS, webhook_url: str | None = None, webhook_events: list[str] | None = None, timeout: int | None = None) -> Operation | None:
        return self._gpu_update(fabric_name, name, "DELETE", servers, mode=mode, webhook_url=webhook_url, webhook_events=webhook_events, timeout=timeout)

    def _gpu_update(self, fabric_name: str, name: str, operation: str, servers: list[str | dict], *, mode: OperationMode, webhook_url: str | None, webhook_events: list[str] | None, timeout: int | None = None) -> Operation | None:
        _require(fabric_name, "fabric_name")
        _require(name, "name")
        if not servers:
            raise ValueError("servers list must not be empty.")
        body = {"operation": operation, "servers": [_normalize_server(s) for s in servers]}
        _attach_webhook_fields(body, mode, webhook_url, webhook_events)
        result = self._transport.patch(f"fabrics/{fabric_name}/tenants/{name}", json_body=body, mode=mode, timeout=timeout)
        if _looks_like_operation(result):
            return Operation.from_submission(result)
        return None


def _require(value: Any, field_name: str) -> None:
    if value is None or (isinstance(value, str) and not value.strip()):
        raise ValueError(f"{field_name} is required and cannot be empty.")


def _validate_max_gpus(value: int) -> None:
    if not isinstance(value, int) or isinstance(value, bool):
        raise ValueError(f"max_gpus_allowed must be an integer; got {type(value).__name__}.")
    if value == -1 or value >= 1:
        return
    raise ValueError(f"max_gpus_allowed must be a positive integer or -1 (unlimited); got {value}.")


def _normalize_server(spec: str | dict) -> dict:
    if isinstance(spec, str):
        return {"serverName": spec, "shared": False}
    if isinstance(spec, dict):
        if "serverName" not in spec:
            raise ValueError("Server spec dict must contain 'serverName'.")
        return {"serverName": spec["serverName"], "shared": bool(spec.get("shared", False))}
    raise TypeError(f"Server entries must be str or dict; got {type(spec).__name__}.")


def _attach_webhook_fields(body: dict, mode: OperationMode, webhook_url: str | None, webhook_events: list[str] | None) -> None:
    if mode is not OperationMode.ASYNC_WEBHOOK:
        if webhook_url is not None or webhook_events is not None:
            raise ValueError("webhook_url and webhook_events are only valid when mode=OperationMode.ASYNC_WEBHOOK.")
        return
    if not webhook_url:
        raise ValueError("webhook_url is required when mode is ASYNC_WEBHOOK.")
    parsed = urlparse(webhook_url)
    if not parsed.scheme or not parsed.netloc:
        raise ValueError(f"webhook_url must be an absolute URL with scheme and host; got {webhook_url!r}.")
    if not webhook_events or not isinstance(webhook_events, list):
        raise ValueError("webhook_events must be a non-empty list of strings.")
    for ev in webhook_events:
        if not isinstance(ev, str) or not ev.strip():
            raise ValueError(f"webhook_events entries must be non-empty strings; got {ev!r}.")
    body["enableWebhook"] = True
    body["webhookUrl"] = webhook_url
    body["webhookEvents"] = list(webhook_events)


def _looks_like_operation(result: Any) -> bool:
    return isinstance(result, dict) and _OPERATION_KEYS.issubset(result.keys())

File: ones-gfx-sdk-go/ones_gfx/resources/tenants.go
package resources

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
	"strings"

	"github.com/aviznetworks/ones-gfx-sdk/ones-gfx-sdk-go/ones_gfx"
)

// TenantsResource handles tenant lifecycle operations.
type TenantsResource struct {
	transport *ones_gfx.Transport
}

// NewTenantsResource constructs a TenantsResource.
func NewTenantsResource(transport *ones_gfx.Transport) *TenantsResource {
	return &TenantsResource{transport: transport}
}

// CreateTenantRequest holds parameters for tenant creation.
type CreateTenantRequest struct {
	Name           string
	Description    string
	MaxGPUsAllowed int
	Shared         bool
}

// ServerSpec describes a GPU server to allocate/deallocate.
type ServerSpec struct {
	ServerName string
	Shared     bool
}

// ServerSpecsFromNames converts server hostnames into ServerSpec entries.
func ServerSpecsFromNames(names []string) []ServerSpec {
	servers := make([]ServerSpec, 0, len(names))
	for _, name := range names {
		servers = append(servers, ServerSpec{ServerName: name})
	}
	return servers
}

// List retrieves all tenants in a fabric.
// Maps to GET /fabrics/{fabricName}/tenants.
func (r *TenantsResource) List(ctx context.Context, fabricName string) ([]ones_gfx.Tenant, error) {
	if fabricName == "" {
		return nil, fmt.Errorf("fabricName is required")
	}

	path := fmt.Sprintf("fabrics/%s/tenants", fabricName)
	body, err := r.transport.Get(path, nil)
	if err != nil {
		return nil, err
	}

	obj, ok := body.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: %T", body)
	}

	tenantsRaw, ok := obj["tenants"]
	if !ok {
		return nil, fmt.Errorf("response missing 'tenants' field")
	}

	tenantsList, ok := tenantsRaw.([]interface{})
	if !ok {
		return nil, fmt.Errorf("'tenants' field is not an array")
	}

	tenants := make([]ones_gfx.Tenant, 0, len(tenantsList))
	for _, item := range tenantsList {
		jsonBytes, err := remarshal(item)
		if err != nil {
			continue
		}
		var tenant ones_gfx.Tenant
		if err := unmarshalJSON(jsonBytes, &tenant); err != nil {
			continue
		}
		tenants = append(tenants, tenant)
	}

	return tenants, nil
}

// Get retrieves a single tenant by name.
// Maps to GET /fabrics/{fabricName}/tenants/{tenantName}.
func (r *TenantsResource) Get(ctx context.Context, fabricName, tenantName string) (*ones_gfx.Tenant, error) {
	if fabricName == "" {
		return nil, fmt.Errorf("fabricName is required")
	}
	if tenantName == "" {
		return nil, fmt.Errorf("tenantName is required")
	}

	path := fmt.Sprintf("fabrics/%s/tenants/%s", fabricName, tenantName)
	body, err := r.transport.Get(path, nil)
	if err != nil {
		return nil, err
	}

	// Response wraps tenant in {"tenant": {...}}
	obj, ok := body.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: %T", body)
	}

	tenantData, ok := obj["tenant"]
	if !ok {
		tenantData = body
	}

	jsonBytes, err := remarshal(tenantData)
	if err != nil {
		return nil, err
	}

	var tenant ones_gfx.Tenant
	if err := unmarshalJSON(jsonBytes, &tenant); err != nil {
		return nil, err
	}

	return &tenant, nil
}

// AvailableServers lists GPU servers available for allocation.
// Maps to GET /fabrics/{fabricName}/available_servers.
func (r *TenantsResource) AvailableServers(ctx context.Context, fabricName string) ([]string, error) {
	if fabricName == "" {
		return nil, fmt.Errorf("fabricName is required")
	}

	path := fmt.Sprintf("fabrics/%s/available_servers", fabricName)
	body, err := r.transport.Get(path, nil)
	if err != nil {
		return nil, err
	}

	obj, ok := body.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected response type: %T", body)
	}

	serversRaw, ok := obj["availableGPUs"]
	if !ok {
		return []string{}, nil
	}

	serversList, ok := serversRaw.([]interface{})
	if !ok {
		return []string{}, nil
	}

	servers := make([]string, 0, len(serversList))
	for _, s := range serversList {
		if str, ok := s.(string); ok {
			servers = append(servers, str)
		}
	}

	return servers, nil
}

// Create creates a tenant (synchronous mode).
// Maps to POST /fabrics/{fabricName}/tenants.
func (r *TenantsResource) Create(ctx context.Context, fabricName string, req CreateTenantRequest, opts ...ones_gfx.CallOption) (*ones_gfx.Tenant, error) {
	if err := validateCreateRequest(fabricName, req); err != nil {
		return nil, err
	}

	timeout, webhookURL, webhookEvents := ones_gfx.ApplyCallOptions(opts...)
	if webhookURL != "" || len(webhookEvents) > 0 {
		return nil, fmt.Errorf("webhook options are only valid for async methods")
	}
	body := buildCreateBody(req)
	path := fmt.Sprintf("fabrics/%s/tenants", fabricName)

	result, err := r.transport.Post(path, body, ones_gfx.OperationModeSynchronous, timeout)
	if err != nil {
		return nil, err
	}

	return extractTenant(result)
}

// CreateAsync creates a tenant (async mode).
// Returns an Operation handle for polling or webhook delivery.
func (r *TenantsResource) CreateAsync(ctx context.Context, fabricName string, req CreateTenantRequest, opts ...ones_gfx.CallOption) (*ones_gfx.Operation, error) {
	if err := validateCreateRequest(fabricName, req); err != nil {
		return nil, err
	}

	timeout, webhookURL, webhookEvents := ones_gfx.ApplyCallOptions(opts...)
	body := buildCreateBody(req)
	if err := attachWebhookFields(body, webhookURL, webhookEvents); err != nil {
		return nil, err
	}

	path := fmt.Sprintf("fabrics/%s/tenants", fabricName)
	mode := ones_gfx.OperationModeAsyncPoll
	if webhookURL != "" {
		mode = ones_gfx.OperationModeAsyncWebhook
	}

	result, err := r.transport.Post(path, body, mode, timeout)
	if err != nil {
		return nil, err
	}

	return extractOperation(result)
}

// Delete deletes a tenant (synchronous mode).
// Maps to DELETE /fabrics/{fabricName}/tenants/{tenantName}.
func (r *TenantsResource) Delete(ctx context.Context, fabricName, tenantName string, opts ...ones_gfx.CallOption) error {
	if fabricName == "" {
		return fmt.Errorf("fabricName is required")
	}
	if tenantName == "" {
		return fmt.Errorf("tenantName is required")
	}

	timeout, webhookURL, webhookEvents := ones_gfx.ApplyCallOptions(opts...)
	if webhookURL != "" || len(webhookEvents) > 0 {
		return fmt.Errorf("webhook options are only valid for async methods")
	}
	path := fmt.Sprintf("fabrics/%s/tenants/%s", fabricName, tenantName)

	_, err := r.transport.Delete(path, nil, ones_gfx.OperationModeSynchronous, timeout)
	return err
}

// DeleteAsync deletes a tenant (async mode).
func (r *TenantsResource) DeleteAsync(ctx context.Context, fabricName, tenantName string, opts ...ones_gfx.CallOption) (*ones_gfx.Operation, error) {
	if fabricName == "" {
		return nil, fmt.Errorf("fabricName is required")
	}
	if tenantName == "" {
		return nil, fmt.Errorf("tenantName is required")
	}

	timeout, webhookURL, webhookEvents := ones_gfx.ApplyCallOptions(opts...)
	path := fmt.Sprintf("fabrics/%s/tenants/%s", fabricName, tenantName)

	var body map[string]interface{}
	if webhookURL != "" || len(webhookEvents) > 0 {
		body = make(map[string]interface{})
		if err := attachWebhookFields(body, webhookURL, webhookEvents); err != nil {
			return nil, err
		}
	}

	mode := ones_gfx.OperationModeAsyncPoll
	if webhookURL != "" {
		mode = ones_gfx.OperationModeAsyncWebhook
	}

	result, err := r.transport.Delete(path, body, mode, timeout)
	if err != nil {
		return nil, err
	}

	return extractOperation(result)
}

// AllocateGPUs adds GPU servers to a tenant (synchronous mode).
// Maps to PATCH /fabrics/{fabricName}/tenants/{tenantName}.
func (r *TenantsResource) AllocateGPUs(ctx context.Context, fabricName, tenantName string, servers []ServerSpec, opts ...ones_gfx.CallOption) error {
	return r.gpuUpdate(ctx, fabricName, tenantName, "ADD", servers, opts...)
}

// AllocateGPUsAsync adds GPU servers (async mode).
func (r *TenantsResource) AllocateGPUsAsync(ctx context.Context, fabricName, tenantName string, servers []ServerSpec, opts ...ones_gfx.CallOption) (*ones_gfx.Operation, error) {
	return r.gpuUpdateAsync(ctx, fabricName, tenantName, "ADD", servers, opts...)
}

// DeallocateGPUs removes GPU servers from a tenant (synchronous mode).
func (r *TenantsResource) DeallocateGPUs(ctx context.Context, fabricName, tenantName string, servers []ServerSpec, opts ...ones_gfx.CallOption) error {
	return r.gpuUpdate(ctx, fabricName, tenantName, "DELETE", servers, opts...)
}

// DeallocateGPUsAsync removes GPU servers (async mode).
func (r *TenantsResource) DeallocateGPUsAsync(ctx context.Context, fabricName, tenantName string, servers []ServerSpec, opts ...ones_gfx.CallOption) (*ones_gfx.Operation, error) {
	return r.gpuUpdateAsync(ctx, fabricName, tenantName, "DELETE", servers, opts...)
}

// gpuUpdate handles sync GPU allocate/deallocate.
func (r *TenantsResource) gpuUpdate(ctx context.Context, fabricName, tenantName, operation string, servers []ServerSpec, opts ...ones_gfx.CallOption) error {
	if fabricName == "" {
		return fmt.Errorf("fabricName is required")
	}
	if tenantName == "" {
		return fmt.Errorf("tenantName is required")
	}
	if len(servers) == 0 {
		return fmt.Errorf("servers list cannot be empty")
	}

	timeout, webhookURL, webhookEvents := ones_gfx.ApplyCallOptions(opts...)
	if webhookURL != "" || len(webhookEvents) > 0 {
		return fmt.Errorf("webhook options are only valid for async methods")
	}
	body := buildGPUUpdateBody(operation, servers)
	path := fmt.Sprintf("fabrics/%s/tenants/%s", fabricName, tenantName)

	_, err := r.transport.Patch(path, body, ones_gfx.OperationModeSynchronous, timeout)
	return err
}

// gpuUpdateAsync handles async GPU allocate/deallocate.
func (r *TenantsResource) gpuUpdateAsync(ctx context.Context, fabricName, tenantName, operation string, servers []ServerSpec, opts ...ones_gfx.CallOption) (*ones_gfx.Operation, error) {
	if fabricName == "" {
		return nil, fmt.Errorf("fabricName is required")
	}
	if tenantName == "" {
		return nil, fmt.Errorf("tenantName is required")
	}
	if len(servers) == 0 {
		return nil, fmt.Errorf("servers list cannot be empty")
	}

	timeout, webhookURL, webhookEvents := ones_gfx.ApplyCallOptions(opts...)
	body := buildGPUUpdateBody(operation, servers)
	if err := attachWebhookFields(body, webhookURL, webhookEvents); err != nil {
		return nil, err
	}

	path := fmt.Sprintf("fabrics/%s/tenants/%s", fabricName, tenantName)
	mode := ones_gfx.OperationModeAsyncPoll
	if webhookURL != "" {
		mode = ones_gfx.OperationModeAsyncWebhook
	}

	result, err := r.transport.Patch(path, body, mode, timeout)
	if err != nil {
		return nil, err
	}

	return extractOperation(result)
}

// Helper functions

func validateCreateRequest(fabricName string, req CreateTenantRequest) error {
	if fabricName == "" {
		return fmt.Errorf("fabricName is required")
	}
	if req.Name == "" {
		return fmt.Errorf("tenant name is required")
	}
	if req.MaxGPUsAllowed != ones_gfx.UnlimitedGPUs && req.MaxGPUsAllowed < 1 {
		return fmt.Errorf("maxGPUsAllowed must be -1 (unlimited) or >= 1, got %d", req.MaxGPUsAllowed)
	}
	return nil
}

func buildCreateBody(req CreateTenantRequest) map[string]interface{} {
	return map[string]interface{}{
		"tenantName":     req.Name,
		"description":    req.Description,
		"maxGpusAllowed": req.MaxGPUsAllowed,
		"shared":         req.Shared,
	}
}

func buildGPUUpdateBody(operation string, servers []ServerSpec) map[string]interface{} {
	serverMaps := make([]map[string]interface{}, len(servers))
	for i, s := range servers {
		serverMaps[i] = map[string]interface{}{
			"serverName": s.ServerName,
			"shared":     s.Shared,
		}
	}
	return map[string]interface{}{
		"operation": operation,
		"servers":   serverMaps,
	}
}

func attachWebhookFields(body map[string]interface{}, webhookURL string, webhookEvents []string) error {
	if webhookURL == "" && len(webhookEvents) == 0 {
		return nil
	}
	if webhookURL == "" || len(webhookEvents) == 0 {
		return fmt.Errorf("webhookURL and webhookEvents must be provided together")
	}
	parsed, err := url.Parse(webhookURL)
	if err != nil || parsed.Scheme == "" || parsed.Host == "" {
		return fmt.Errorf("webhookURL must be an absolute URL with scheme and host, got %q", webhookURL)
	}
	for _, event := range webhookEvents {
		if strings.TrimSpace(event) == "" {
			return fmt.Errorf("webhookEvents must contain non-empty strings")
		}
	}
	body["enableWebhook"] = true
	body["webhookUrl"] = webhookURL
	body["webhookEvents"] = webhookEvents
	return nil
}

func extractTenant(result interface{}) (*ones_gfx.Tenant, error) {
	obj, ok := result.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("unexpected result type: %T", result)
	}

	tenantData, ok := obj["tenant"]
	if !ok {
		tenantData = result
	}

	jsonBytes, err := remarshal(tenantData)
	if err != nil {
		return nil, err
	}

	var tenant ones_gfx.Tenant
	if err := unmarshalJSON(jsonBytes, &tenant); err != nil {
		return nil, err
	}

	return &tenant, nil
}

func extractOperation(result interface{}) (*ones_gfx.Operation, error) {
	jsonBytes, err := remarshal(result)
	if err != nil {
		return nil, err
	}

	var op ones_gfx.Operation
	if err := unmarshalJSON(jsonBytes, &op); err != nil {
		return nil, err
	}

	return &op, nil
}

// remarshal converts interface{} back to JSON bytes for unmarshaling into structs.
func remarshal(v interface{}) ([]byte, error) {
	return json.Marshal(v)
}

func unmarshalJSON(data []byte, v interface{}) error {
	return json.Unmarshal(data, v)
}
